### ISSUE TITLE
3.6.0: deleting Dynamic Sprite in "Room Leave" event may make objects invisible during transition

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1970]
User found a visual regression in 3.6.0, which turned out to be a serious issue, uncovering the *not less* serious problem in AGS events design, which I frankly did not pay attention to previously.

### Problem

If you have a room object with a Dynamic Sprite assigned to it, and delete that sprite in "room leave" event, the object will visually disappear during room transition (fade out). That *seem* like a logical thing to happen if you think about it, except **it did not happen** in previous versions of AGS, and obviously nobody will expect that now.

Hence this is a **very critical** regression.

### To reproduce
Make a game with 2 rooms.
In room 1 add a room object, say oObject1.
Use following room script (don't forget to connect room load and room leave events):
```
DynamicSprite *SPR;

function room_Load()
{
    oObject1.Graphic = SPR.Graphic;
}

function room_Leave()
{
    SPR.Delete();
}

function on_key_press(eKeyCode k)
{
    if (k == eKeySpace)
    {
        player.ChangeRoom(2);
    }
}
```
Run the game and press Space. Observe object becomes invisible right before the transition.

### The reasons why this happens

The "design problem" that I mentioned earlier is this. AGS has "room leave" event, which is traditionally used to release all room resources that were created in script. But this event unfortunately takes place *before* the transition. And there's no script event that would trigger *after* the transition, something corresponding to "room load".

Obviously, as the dynamic sprite is deleted, we should expect that it's not drawn anywhere on screen. So if it's deleted *before* the transition, then it might as well be not drawn *during* the transition.

### The reasons why this "worked", and stopped now

As a very common in AGS (sadly), it worked because of some coincidental combination of things... But most importantly: the dedicated object's texture remained in memory! Every game object in AGS has a texture assigned to it, and this texture has a copy of the last used sprite on it.
Prior to 3.6.0 the object textures were only released as the room got unloaded.
In 3.6.0 they are also released for objects that had a DynamicSprite assigned, and that sprite got deleted.

### Workarounds

User may fade out using script commands before deleting the sprites, while having room transition style set to Instant. This is also a common solution for custom transition styles.

### Potential fix

It *seems* that this regression may be fixed by simply not releasing an object's texture when the sprite is deleted. Theoretically, it should keep the old sprite's image for *as long as the object is not updated*, that is - a game logic update is not run. This is a quite a flimsy solution, but my preliminary tests indicate this might work, *for now*. Any further change or optimization to the engine may break it again, so this is something to watch out for, sadly... 

Considering the texture release on its own, it could be a mistake to do upon a sprite deletion overall (not in relevance to this problem). Generally speaking, there's nothing wrong in objects retaining their texture surface for most of the time. Also, if the next assigned sprite will be of the same size, then the last texture surface may as well be reused. So probably it will be a good thing to not release it in this case, and only release when unloading a room.

### The future of this problem

In the long run, I believe, we must add a new room event, which is called right after fadeout (after the room got drawn for the last time) and before this room gets unloaded. Something corresponding to "room load" (so - "room unload"?). This will be the correct place to release dynamic sprites and other resources.

For backwards compatible branch, we will have to basically make sure that the object's texture is *never updated during the fadeout*...
There may be other potential workarounds here, but the ones that come to mind are either "ugly hacks", or redesign of how DynamicSprite, or object's view works.

For example, a DynamicSprite.Delete may not actually delete a sprite immediately, but mark it as "free", while the sprite itself will be kept so long as its assigned to anything. In other words: assigning it by a Graphic index would also increment its reference count. Maybe this is how it should be working from the start in AGS...

Another approach would be to consider objects to have their own separate image *logically*, that is retained, keeping last content, if no sprite is assigned to the object (so there has to be a special "no sprite" value in properties). Although, tbh, this sounds "incorrect".

All the above is a question of the engine design, not a quick fix though.

**Comment by fernewelten** [comment]
The situation is even messier because one of the room transitions is Crossfade. In that case, the new screen is already showing before visuals from the old screens can be released. 

We _need_ an event where dynamic sprites etc. can be safely released. The clean design would be a dedicated event `room_Unload` that would fire __after__ the room transition, whatever it may be, with the restriction that this event must not spend time (similarly to the `room_Load` event). i.e., must not Say or Wait etc.

**Comment by ivan-mogilko** [comment]
> The situation is even messier because one of the room transitions is Crossfade. In that case, the new screen is already showing before visuals from the old room can be released.

If I remember correctly, the crossfade uses the last room's "screenshot". It cannot draw previous room and new room at the same time, because only 1 room is in memory at any time.

**Comment by ivan-mogilko** [comment]
In regards to the possible fix:
> It seems that this regression may be fixed by simply not releasing an object's texture when the sprite is deleted.

Have to be very careful with this, because:
7ed51861898d864902b0f61b0cb7d0acadef1fe3

Possibly, the texture itself may be kept, but the SpriteID should be reset.

**Comment by ivan-mogilko** [comment]
Should be fixed by 38d0dd3d7f0d6a4624d1f908f2e8d3c1abfee52a

@fernewelten , ironically, it appears that Crossfade transition did not have same issue, probably the screenshot is taken using old render state.

**Comment by ivan-mogilko** [comment]
Duh! keep fixing mistakes in the same area.

First of all, I found that all this time the wrong texture cache has been cleared for characters: 144ac42f99c6be0a1368c74e9a9baaf7440dab70
**This** was the reason why characters did not become invisible during fade-out (as per the bug described in this ticket) while objects were.

On another hand I found that this fix I did here was breaking another thing, or rather contributing to the oversight in texture cache implementation. Which is now fixed too (hopefully): cc109132340c96e186e8b9a24bc908b3271bb5f6
